File: src/finance_mask/scanner/column_matcher.py
"""列头匹配引擎"""
from __future__ import annotations

import json
import re
from pathlib import Path
from typing import Optional

from ..models.strategy import ColumnRule, MatchType
from ..models.site import DetectedType, ActionType

# 默认配置文件路径
DEFAULT_CONFIG_PATH = Path(__file__).parent.parent.parent.parent / "config" / "column_rules.json"


def _load_rules_from_config(config_path: Path) -> list[ColumnRule]:
    """从配置文件加载列头规则"""
    if not config_path.exists():
        raise FileNotFoundError(f"配置文件不存在: {config_path}")
    
    with open(config_path, 'r', encoding='utf-8') as f:
        config = json.load(f)
    
    # 类型映射
    match_type_map = {
        "exact": MatchType.EXACT,
        "regex": MatchType.REGEX,
        "position": MatchType.POSITION,
    }
    
    action_type_map = {
        "precision": ActionType.PRECISION,
        "perturb": ActionType.PERTURB,
        "mask": ActionType.MASK,
        "alias": ActionType.ALIAS,
        "mask_name": ActionType.MASK_NAME,
        "mask_account": ActionType.MASK_ACCOUNT,
        "differential_shift": ActionType.DIFFERENTIAL_SHIFT,
        "proportional_scale": ActionType.PROPORTIONAL_SCALE,
    }
    
    detected_type_map = {
        "amount": DetectedType.AMOUNT,
        "entity": DetectedType.ENTITY,
        "person": DetectedType.PERSON,
        "account": DetectedType.ACCOUNT,
    }
    
    rules = []
    for rule_data in config.get("rules", []):
        rule = ColumnRule(
            match_type=match_type_map[rule_data["match_type"]],
            pattern=rule_data["pattern"],
            action=action_type_map[rule_data["action"]],
            params=rule_data.get("params"),
            detected_type=detected_type_map[rule_data["detected_type"]],
            priority=rule_data.get("priority", 0),
        )
        rules.append(rule)
    
    return rules


class ColumnMatcher:
    """列头匹配引擎"""

    def __init__(self, custom_rules: Optional[list[ColumnRule]] = None, config_path: Optional[Path] = None):
        self._rules: list[ColumnRule] = _load_rules_from_config(config_path or DEFAULT_CONFIG_PATH)
        if custom_rules:
            self._rules.extend(custom_rules)
        # 按优先级排序
        self._rules.sort(key=lambda r: r.priority)

    def match(self, header: str) -> Optional[ColumnRule]:
        """
        匹配列名，返回匹配的 ColumnRule 或 None
        
        匹配顺序：精确匹配 > 正则匹配
        """
        if not header or not isinstance(header, str):
            return None

        header = header.strip()
        if not header:
            return None

        # 第一轮：精确匹配
        for rule in self._rules:
            if rule.match_type == MatchType.EXACT:
                if rule.pattern == header:
                    return rule

        # 第二轮：正则匹配
        for rule in self._rules:
            if rule.match_type == MatchType.REGEX:
                if re.search(rule.pattern, header) is not None:
                    return rule

        return None

    def add_rule(self, rule: ColumnRule) -> None:
        """添加自定义规则"""
        self._rules.append(rule)
        self._rules.sort(key=lambda r: r.priority)

File: rust/src/column_matcher.rs
//! 列头匹配引擎 —— 移植自 src/finance_mask/scanner/column_matcher.py（104 行）。
//!
//! 匹配行为以 Python 为准，仅两种方式：
//! - Exact：字符串相等
//! - Regex：`re.search` 语义（fancy-regex `is_match`）
//!
//! `MatchType::Position` 在 Python 侧从未实现匹配（仅加载不生效，无调用方），
//! 不参与匹配（2026-09-06 裁决）。

use crate::models::{ColumnRule, MatchType};

/// 列头匹配引擎（对应 column_matcher.py ColumnMatcher）
#[derive(Debug)]
pub struct ColumnMatcher {
    rules: Vec<ColumnRule>,
}

impl ColumnMatcher {
    /// 以给定规则构建；按 priority 升序稳定排序
    /// （数字越小优先级越高，同优先级保持传入顺序，对应 Python `sort(key=priority)`）
    pub fn new(rules: Vec<ColumnRule>) -> Self {
        let mut rules = rules;
        rules.sort_by_key(|r| r.priority);
        Self { rules }
    }

    /// 匹配列名，返回命中的 ColumnRule 或 None。
    ///
    /// 匹配顺序：精确匹配 > 正则匹配（两轮次序，priority 只在轮内生效）
    pub fn match_header(&self, header: &str) -> Option<&ColumnRule> {
        let header = header.trim();
        if header.is_empty() {
            return None;
        }

        // 第一轮：精确匹配
        for rule in &self.rules {
            if rule.match_type == MatchType::Exact && rule.pattern == header {
                return Some(rule);
            }
        }

        // 第二轮：正则匹配
        for rule in &self.rules {
            if rule.match_type == MatchType::Regex {
                let re = fancy_regex::Regex::new(&rule.pattern).unwrap_or_else(|e| {
                    panic!("列头规则正则编译失败 (pattern={:?}): {e}", rule.pattern)
                });
                match re.is_match(header) {
                    Ok(true) => return Some(rule),
                    Ok(false) => {}
                    // 对应 Python re.search 的运行时异常：崩溃而非静默漏配
                    Err(e) => panic!("列头规则正则匹配失败 (pattern={:?}): {e}", rule.pattern),
                }
            }
        }

        None
    }
}
